vms/proposervm/vm.go

```diff
@@ -315,13 +300,83 @@ func (vm *VM) SetPreference(ctx context.Context, preferred ids.ID) error {
 		return vm.ChainVM.SetPreference(ctx, preferred)
 	}
 
-	if err := vm.ChainVM.SetPreference(ctx, blk.getInnerBlk().ID()); err != nil {
+	innerBlkID := blk.getInnerBlk().ID()
+	if err := vm.ChainVM.SetPreference(ctx, innerBlkID); err != nil {
 		return err
 	}
 
+	vm.ctx.Log.Debug("set preference",
+		zap.Stringer("blkID", preferred),
+		zap.Stringer("innerBlkID", innerBlkID),
+	)
+	return nil
+}
+
+func (vm *VM) WaitForEvent(ctx context.Context) (common.Message, error) {
+	for {
+		if err := ctx.Err(); err != nil {
+			vm.ctx.Log.Debug("Aborting WaitForEvent, context is done", zap.Error(err))
+			return 0, err
+		}
+
+		timeToBuild, shouldWait, err := vm.timeToBuild(ctx)
+		if err != nil {
+			vm.ctx.Log.Debug("Aborting WaitForEvent", zap.Error(err))
+			return 0, err
+		}
+
+		// If we are pre-fork or haven't finished bootstrapping yet, we should
+		// directly forward the inner VM's events.
+		if !shouldWait {
+			vm.ctx.Log.Debug("Waiting for inner VM event (pre-fork)")
```

Comment:
This code path is also used for state-sync, not just when the VM is pre-fork.

Reply:
[Fixed](https://github.com/ava-labs/avalanchego/pull/3999/commits/aa40197de464a9646a64b79fa21ff08b7d3b8db7#diff-83207c67653d884434d01f4c9ad5255cdda73dabef39063346ac96f4a039530fR331-R337).